src/perturbations.jl

```diff
@@ -12,49 +12,92 @@ struct Hierarchy{T<:Real, PI<:PerturbationIntegrator, CP<:AbstractCosmoParams{T}
     Θᵖ = OffsetVector(view(u, (ℓᵧ+2):(2ℓᵧ+2)), 0:ℓᵧ)  # indexed 0 through ℓᵧ
-    𝒩 = OffsetVector(view(u, (2(ℓᵧ+1) + 1):(2(ℓᵧ+1)+(ℓ_ν+1))) , 0:ℓ_ν)  # indexed 0 through ℓ_ν
-    Φ, δ, v, δ_b, v_b = view(u, (2(ℓᵧ+1)+(ℓ_ν+1)+1):(2(ℓᵧ+1)+(ℓ_ν+1)+5)) #getting a little messy...
-    return Θ, Θᵖ, 𝒩, Φ, δ, v, δ_b, v_b
+    𝒩 = OffsetVector(view(u, (2(ℓᵧ+1) + 1):(2(ℓᵧ+1)+ℓ_ν+1)) , 0:ℓ_ν)  # indexed 0 through ℓ_ν
+    ℳ = OffsetVector(view(u, (2(ℓᵧ+1)+(ℓ_ν+1)+1):(2(ℓᵧ+1)+(ℓ_ν+1)+(ℓ_mν+1)*nq )) , 0:(ℓ_mν+1)*nq -1)  # indexed 0 through ℓ_mν
+    Φ, δ, v, δ_b, v_b = view(u, ((2(ℓᵧ+1)+(ℓ_ν+1)+(ℓ_mν+1)*nq)+1 :(2(ℓᵧ+1)+(ℓ_ν+1)+(ℓ_mν+1)*nq)+5)) #getting a little messy...
+    return Θ, Θᵖ, 𝒩, ℳ, Φ, δ, v, δ_b, v_b
+end
+
+function ρ_σ(ℳ0,ℳ2,bg,a,par::AbstractCosmoParams) #a mess
+    #Do q integrals to get the massive neutrino metric perturbations
+    #MB eqn (55)
+    Tν =  (par.N_ν/3)^(1/4) *(4/11)^(1/3) * (15/ π^2 *ρ_crit(par) *par.Ω_r)^(1/4)
+    logqmin,logqmax=log10(Tν/30),log10(Tν*30)#1e-6,1e-1
+
+    #FIXME: avoid repeating code? and maybe put general integrals in utils?
+    m = par.Σm_ν
+    nq = length(ℳ0) #assume we got this right
+    ϵx(x, am) = √(xq2q(x,logqmin,logqmax)^2 + (am)^2)
+    Iρ(x) = xq2q(x,logqmin,logqmax)^2  * ϵx(x, a*m) * f0(xq2q(x,logqmin,logqmax),par) / dxdq(xq2q(x,logqmin,logqmax),logqmin,logqmax)
+    Iσ(x) = xq2q(x,logqmin,logqmax)^2  * (xq2q(x,logqmin,logqmax)^2 /ϵx(x, a*m)) * f0(xq2q(x,logqmin,logqmax),par) / dxdq(xq2q(x,logqmin,logqmax),logqmin,logqmax)
+
+    xq,wq = bg.quad_pts,bg.quad_wts
+    ρ = 4π*sum(Iρ.(xq).*ℳ0.*wq)
+    σ = 4π*sum(Iσ.(xq).*ℳ2.*wq)
+    # #a-dependence has been moved into Einstein eqns, as have consts in σ
+    return ρ,σ
 end
 
 # BasicNewtonian comes from Callin+06 and the Dodelson textbook (dispatches on hierarchy.integrator)
 function hierarchy!(du, u, hierarchy::Hierarchy{T, BasicNewtonian}, x) where T
     # compute cosmological quantities at time x, and do some unpacking
-    k, ℓᵧ, par, bg, ih = hierarchy.k, hierarchy.ℓᵧ, hierarchy.par, hierarchy.bg, hierarchy.ih
-    Ω_r, Ω_b, Ω_m, N_ν, H₀² = par.Ω_r, par.Ω_b, par.Ω_m, par.N_ν, bg.H₀^2 #add N_ν≡N_eff
+    k, ℓᵧ, par, bg, ih, nq = hierarchy.k, hierarchy.ℓᵧ, hierarchy.par, hierarchy.bg, hierarchy.ih,hierarchy.nq
+    Tν =  (par.N_ν/3)^(1/4) *(4/11)^(1/3) * (15/ π^2 *ρ_crit(par) *par.Ω_r)^(1/4)
+    logqmin,logqmax=log10(Tν/30),log10(Tν*30)
+    q_pts = xq2q.(bg.quad_pts,logqmin,logqmax)
+    Ω_r, Ω_b, Ω_m, N_ν, m_ν, H₀² = par.Ω_r, par.Ω_b, par.Ω_m, par.N_ν, par.Σm_ν, bg.H₀^2 #add N_ν≡N_eff
     ℋₓ, ℋₓ′, ηₓ, τₓ′, τₓ′′ = bg.ℋ(x), bg.ℋ′(x), bg.η(x), ih.τ′(x), ih.τ′′(x)
     a = x2a(x)
     R = 4Ω_r / (3Ω_b * a)
-    Ω_ν =  7N_ν/8 *(4/11)^(4/3) *Ω_r
-    ℓ_ν = 10 #again, for now
+    Ω_ν =  7*(2/3)*N_ν/8 *(4/11)^(4/3) *Ω_r
+    ℓ_ν = 10 #again, for now - should this be higher??
+    ℓ_mν =  ℓ_ν #come back to put ℓmaxs in hierarchy
+    norm𝒩′ = 1.0 /(Ω_ν * bg.ρ_crit / 2)# par.N_ν) #Normalization to match 𝒩 after integrating, par.N_ν->2
```

Comment:
Where does this factor of 2 come from?

Reply:
The neutrino perturbation \scrM is normalized to the neutrino perturbation \scrN. However, the contribution to the metric eqn should be smaller by a factor of 2 since there are 2 massless neutrinos and one massive one (w/ equal Neff sharing). The two cancels the 2 in the denominator of 2/3 in the expression for \Omega_\nu 3 lines up.